internal/ingest/segmented_sync.go

```diff
@@ -0,0 +1,194 @@
+		} else {
+			nextDepthLimit = remainingDepth
+		}
+	}
+}
+
+func (ing *Ingester) syncAdvertisements(ctx context.Context, pid peer.ID, peerAddr multiaddr.Multiaddr, maxDepth int, stopAt ipld.Link, hookFunc legs.BlockHookFunc) (cid.Cid, error) {
+
+	// If segmented sync is disabled, sync the ad in one go according to the given max depth
+	nextDepthLimit := ing.cfg.SyncSegmentDepthLimit
+	if nextDepthLimit == -1 {
+		log.Debug("segmented sync is disabled; syncing ad in one call")
+		rLimit := recursionLimit(maxDepth)
+		ing.sub.Sync(ctx, pid, cid.Undef, legs.ExploreRecursiveWithStopNode(rLimit, adSequence.Node(), stopAt), peerAddr, legs.AlwaysUpdateLatest(), legs.ScopedBlockHook(hookFunc))
+	}
+
+	// If the configured total max depth is not unlimited and is less than next segment depth limit
+	// then use the configured total depth limit to avoid overshooting sync depth.
+	if maxDepth > 0 && maxDepth < ing.cfg.SyncSegmentDepthLimit {
+		nextDepthLimit = maxDepth
+	}
+
+	// Use blockHook to learn a synced entry chunk's "Next" link.
+	var prevAdLink *ipld.Link
+	var latestLoadErr error
+	opts := []legs.SyncOption{
+		legs.AlwaysUpdateLatest(),
+		legs.ScopedBlockHook(func(pid peer.ID, c cid.Cid) {
+			// Load the ad to learn its previous link
+			// Note that loading is done here because we cannot be sure that the ad will remain in
+			// datastore once the segment sync is done.
+			var ad schema.Advertisement
+			ad, latestLoadErr = ing.loadAd(c)
+			if latestLoadErr == nil {
+				prevAdLink = ad.PreviousID
+			}
+			if hookFunc != nil {
+				hookFunc(pid, c)
+			}
+		}),
+	}
+
+	// Stat sync from undef so that head is queried automatically.
+	nextSegmentCid := cid.Undef
+	// Remember the CID returned by first segment sync as the head CID that is finally returned.
+	headCid := cid.Undef
+	var depthSoFar int
+	for {
+		sel := legs.ExploreRecursiveWithStopNode(selector.RecursionLimitDepth(int64(nextDepthLimit)), adSequence.Node(), stopAt)
+		c, err := ing.sub.Sync(ctx, pid, nextSegmentCid, sel, peerAddr, opts...)
+		if err != nil {
+			log.Errorw("Failed to sync with provider", "err", err)
+			return cid.Undef, err
+		}
+		depthSoFar += nextDepthLimit
+
+		// If returned cid is undef, then there must have been no head, meaning nothing to sync.
+		if c == cid.Undef {
+			return cid.Undef, nil
+		}
+
+		// Set the head cid as the cid returned by the very first segment sync, since the first
+		// segment sync uses cid.Undef and forces head lookup internally.
+		if headCid == cid.Undef {
+			headCid = c
+		}
+
+		// If the latest ad loading failed in block hook then return it here.
+		if latestLoadErr != nil {
+			log.Error("failed to load ad during segmented sync", "err", err)
+			return cid.Undef, latestLoadErr
+		}
+
+		// If there is no previous ad link we are at the end of the ad chain; nothing else to do.
+		if prevAdLink == nil {
+			return headCid, nil
+		}
+		nextSegmentCid = (*(prevAdLink)).(cidlink.Link).Cid
```

Comment:
catch panics

Reply:
Sorry could you helm me understand what could cause a panic here?

Reply:
if the ipld.link was not a cidlink.Link. I think we can probably ignore that edge case though.

Reply:
There are a lot of places in the codebase where we cast `ipld.link` to `codlink.Link`. If this is not a blocker just now, I'll defer safety check in another PR and apply it across the entire repo.

Reply:
yeah, low priority. can be safely ignored for now, i think

Reply:
https://github.com/filecoin-project/storetheindex/issues/491